### PR TITLE
Add SAS lexer and accompanying files

### PR DESCRIPTION
**Pull request by tomsutch**
Add lexer for SAS.

It's a bit of an odd language with lots of keywords, most of them specific to individual procedures, so I hope I have gone about this in an appropriate way. I haven't attempted to implement all of the procedure-specific stuff by any means.

Happy to change in response to feedback :-)

### CODE REVIEW

**Comment by pyrmont** [comment]
Forgive me, @tomsutch. It might take a bit of time to get through this lexer—there sure are a lot of keywords!

File: lib/rouge/lexers/sas.rb [hunk 1]
```diff
@@ -0,0 +1,568 @@
+# -*- coding: utf-8 -*- #
+
+module Rouge
+  module Lexers
+    class SAS < RegexLexer
+      title "SAS"
+      desc "SAS (Statistical Analysis Software)"
+      tag 'sas'
+      filenames '*.sas'
+      mimetypes *%w(
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: We're taking a look at reducing the resource usage of Rouge. To that end, could you pass string literals to `mimetypes`?

File: lib/rouge/lexers/sas.rb [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+        application/x-sas
+        application/x-stat-sas
+        application/x-sas-syntax
+      )
+
+      def self.data_step_statements
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: This is more of a stylistic thing but for consistency, could you name the method with the same name as the instance variable? Ditto for the other methods that wrap reserved words.

File: lib/rouge/lexers/sas.rb [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+          &SYSPBUFF &SYSPRINTTOLIST &SYSPRINTTOLOG &SYSPROCESSID &SYSPROCESSMODE
+          &SYSPROCESSNAME &SYSPROCNAME &SYSRC &SYSSCP &SYSSCPL &SYSSITE &SYSSIZEOFLONG
+          &SYSSIZEOFPTR &SYSSIZEOFUNICODE &SYSSTARTID &SYSSTARTNAME &SYSTCPIPHOSTNAME
+          &SYSTIME &SYSTIMEZONE &SYSTIMEZONEIDENT &SYSTIMEZONEOFFSET &SYSUSERID &SYSVER
+          &SYSVLONG &SYSVLONG4 &SYSWARNINGTEXT
+        )
+      end
+
+      # Create a hash with keywords for common PROCs, keyed by PROC name
+      PROC_KEYWORDS = Hash.new
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: Can you put these assignments behind a method wrapper? This avoids the creation of the string arrays unless and until this lexer is used.

File: lib/rouge/lexers/sas.rb [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+          TRANSREG TRANTAB TREE TSCSREG TTEST UCM UNIVARIATE UPLOAD VARCLUS
+          VARCOMP VARIOGRAM VARMAX VASMP X11 X12 X13 XSL
+        )
+      end
+
+      state :basics do
+        # Rules to be parsed before the keywords (which are different depending
+        # on the context)
+
+        rule /\s+/m, Text
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: This would be more correctly `Text::Whitespace`.

File: lib/rouge/lexers/sas.rb [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+        rule /'[0-9a-z]+?'d/i, Literal::Date
+        rule /'.+?'dt/i, Literal::Date
+        rule /'[0-9:]+?([a|p]m)?'t/i, Literal::Date
+
+        rule /'/, Str::Single, :single_string
+        rule /"/, Str::Double, :double_string
+        rule /&[a-z0-9_&.]+/i, Name::Variable
+
+        # numeric constants (Language Reference p91)
+        rule /\d[0-9a-f]*x/i, Num::Hex
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: Hexadecimal numbers couldn't start with a letter? And can you add some examples of these to your visual sample?

File: lib/rouge/lexers/sas.rb [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+        # on the context)
+
+        rule /\s+/m, Text
+
+        # Single-line comments (between * and ;) - these can actually go onto multiple lines
+        # case 1 - where it starts a line
+        rule /^\s*%?\*[^;]*;/m, Comment::Single
+        # case 2 - where it follows the previous statement on the line (after a semicolon)
+        rule /(;)(\s*)(%?\*[^;]*;)/m do
+          groups Punctuation, Text, Comment::Single
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: Same comment here about `Text::Whitespace`.

File: lib/rouge/lexers/sas.rb [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+
+        # auto variables from DATA step (Language Reference p46, p37)
+        rule /\b(_n_|_error_|_file_|_infile_|_msg_|_iorc_|_cmd_)\b/i, Name::Builtin::Pseudo
+
+        # auto variable list names
+        rule /\b(_character_|_numeric_|_all_)\b/i, Name::Builtin
+
+        # datalines/cards etc
+        rule /\b(datalines|cards)(\s*)(;)/i do
+          groups Keyword, Text, Punctuation
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: `Text::Whitespace` correction.

File: lib/rouge/lexers/sas.rb [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+          push :datalines
+        end
+        rule /\b(datalines4|cards4)(\s*)(;)/i do
+          groups Keyword, Text, Punctuation
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: `Text::Whitespace` correction.

File: lib/rouge/lexers/sas.rb [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+          push :datalines4
+        end
+
+
+        # operators (Language Reference p96)
+        rule %r(\*\*|[\*/\+-]), Operator
+        rule /[^¬~]?=:?|[<>]=?:?/, Operator
+        rule /\b(eq|ne|gt|lt|ge|le|in)\b/i, Operator::Word
+        rule /[&|!¦¬∘~]/, Operator
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: Are these symbols all correct?

File: lib/rouge/lexers/sas.rb [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+        rule /\b(of)\b/i, Operator::Word
+        rule /\b(like)\b/i, Operator::Word # Language Ref p181
+      
+        rule /\d+/, Num::Integer
+
+        rule /\$/, Keyword::Type
+
+        # Macro definitions
+        rule /(%macro|%mend)(\s*)([\w\d]+)/i do
+          groups Keyword, Text, Name::Function
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: `Text::Whitespace` correction.

File: lib/rouge/lexers/sas.rb [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+      end
+
+
+
+      state :root do
+        mixin :basics
+
+        # PROC definitions
+        rule %r((proc)(\s+)(\w+))ix do |m|
+          @proc_name = m[0].split(' ')[1].downcase
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: This should be a local variable rather than an instance variable, right?

File: spec/visual/samples/sas [hunk 12]
```diff
@@ -0,0 +1,71 @@
+* Example SAS script showing syntax highlighting;
+data population;
+/* 
+The first 10 lines of 2017 population data. Source: World Bank
+https://data.worldbank.org/indicator/sp.pop.totl
+Licence: CC BY-4.0
```
Comment (pyrmont) [inline spec/visual/samples/sas]: Is this the correct format for the attribution?

File: lib/rouge/lexers/sas.rb [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+
+        # Missing values (Language Reference p81)
+        rule /\s\.[;\s]/, Keyword::Constant # missing
+        rule /\s\.[a-z_]/, Name::Constant # user-defined missing
+
+        rule /[\(\),;:\{\}\[\]\\\.]/, Punctuation
+
+        rule /@/, Str::Symbol # line hold specifiers
+        rule /\?/, Str::Symbol # used for format modifiers
+
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: This will mean errors in the syntax won't be highlighted. Is that intentional?

File: lib/rouge/lexers/sas.rb [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+        rule /[\(\),;:\{\}\[\]\\\.]/, Punctuation
+
+        rule /@/, Str::Symbol # line hold specifiers
+        rule /\?/, Str::Symbol # used for format modifiers
+        
+        rule /.*/, Text # Fallback for anything we haven't matched so far
+      end
+
+
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: Can you delete this unnecessary whitespace?

File: lib/rouge/lexers/sas.rb [hunk 15]
```diff
@@ -0,0 +1,568 @@
...
+          elsif PROC_KEYWORDS.has_key?(@proc_name) and PROC_KEYWORDS[@proc_name].include? m[0].upcase
+            token Keyword
+          else
+            token Name
+          end
+        end
+
+         mixin :basics2
+      end
+
```
Comment (pyrmont) [inline lib/rouge/lexers/sas.rb]: More unnecessary whitespace.

File: spec/visual/samples/sas [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+run;
+
+* Fit a regression model;
+proc reg data=sashelp.cars;
+    model msrp = enginesize horsepower mpg_city;
+run;
+
+* Plot Fisher's Iris data;
+proc sgplot data=sashelp.iris;
+    scatter x = SepalLength y = SepalWidth / group = Species;
```
Comment (pyrmont) [inline spec/visual/samples/sas]: Are these supposed to be on the same line?

**Comment by pyrmont** [comment]
@tomsutch I do wonder if there's some shortcut to identifying keywords (such as by position) as it seems like there's a lot of matching being done to produce relatively few distinct tokens.

**Comment by tomsutch** [comment]
@pyrmont Thanks for your feedback. I'll take a look at your comments in detail over the coming days.

File: lib/rouge/lexers/sas.rb (hunk 13 above)
Comment (jneen) [reply lib/rouge/lexers/sas.rb]: Even worse, if there's an error, we this will simply render the entire rest of the stream as text...

Comment (jneen) [reply lib/rouge/lexers/sas.rb]: (actually the rest of the line - `.` doesn't match newlines since there's no `/m`)

File: lib/rouge/lexers/sas.rb [hunk 17]
```diff
@@ -0,0 +1,568 @@
...
+      mimetypes *%w(
+        application/x-sas
+        application/x-stat-sas
+        application/x-sas-syntax
+      )
+
+      def self.data_step_statements
+        # from Data step statements - SAS 9.4 Statements reference
+        # http://support.sas.com/documentation/cdl/en/lestmtsref/68024/PDF/default/lestmtsref.pdf
+        @keywords ||= %w(
```
Comment (jneen) [inline lib/rouge/lexers/sas.rb]: We generally use a `Set` for these things for faster lookup.

``` ruby
@data_step_statements ||= Set.new %w(
...
```

should work fine.

File: lib/rouge/lexers/sas.rb (hunk 13 above)
Comment (pyrmont) [reply lib/rouge/lexers/sas.rb]: I of course meant to attach this comment to line 473. @jneen's comments are about the same line. Sorry for the confusion @tomsutch!

File: lib/rouge/lexers/sas.rb (hunk 4 above)
Comment (pyrmont) [reply lib/rouge/lexers/sas.rb]: Ah, my bad. Sorry, @tomsutch!

File: lib/rouge/lexers/sas.rb (hunk 17 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks, have done as part of https://github.com/rouge-ruby/rouge/pull/1107/commits/01c2cc4ec3f09e7ffb2f2ce7571f6324032d052f

File: lib/rouge/lexers/sas.rb (hunk 14 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks; done as part of https://github.com/rouge-ruby/rouge/pull/1107/commits/3191ee40b262959258188a1c4661249f0daf3520

File: lib/rouge/lexers/sas.rb (hunk 5 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Yes, SAS requires a digit at the beginning, to distinguish the number from a variable name - so to write the decimal number 10 in hex you have to say `0ax`. I've added an example in https://github.com/rouge-ruby/rouge/pull/1107/commits/3191ee40b262959258188a1c4661249f0daf3520

File: lib/rouge/lexers/sas.rb (hunk 1 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks; done in https://github.com/rouge-ruby/rouge/pull/1107/commits/3191ee40b262959258188a1c4661249f0daf3520.

File: lib/rouge/lexers/sas.rb (hunk 2 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks; done in https://github.com/rouge-ruby/rouge/pull/1107/commits/01c2cc4ec3f09e7ffb2f2ce7571f6324032d052f.

File: lib/rouge/lexers/sas.rb (hunk 3 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks; done in https://github.com/rouge-ruby/rouge/pull/1107/commits/01c2cc4ec3f09e7ffb2f2ce7571f6324032d052f

File: lib/rouge/lexers/sas.rb [hunk 18]
```diff
@@ -0,0 +1,563 @@
...
+          &SYSPROCESSNAME &SYSPROCNAME &SYSRC &SYSSCP &SYSSCPL &SYSSITE &SYSSIZEOFLONG
+          &SYSSIZEOFPTR &SYSSIZEOFUNICODE &SYSSTARTID &SYSSTARTNAME &SYSTCPIPHOSTNAME
+          &SYSTIME &SYSTIMEZONE &SYSTIMEZONEIDENT &SYSTIMEZONEOFFSET &SYSUSERID &SYSVER
+          &SYSVLONG &SYSVLONG4 &SYSWARNINGTEXT
+        )
+      end
+
+      def self.proc_keywords 
+        # Create a hash with keywords for common PROCs, keyed by PROC name
+        @proc_keywords = Hash.new
```
Comment (ashmaroli) [inline lib/rouge/lexers/sas.rb]: @tomsutch What this does is to assign a new Hash on every call to `:proc_keywords`. Perhaps you meant to use `||=` here as well..:
```suggestion
        @proc_keywords ||= {}
```

File: lib/rouge/lexers/sas.rb (hunk 13 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks both. I have changed to match to the end of the word, rather than to the end of the line, in https://github.com/rouge-ruby/rouge/pull/1107/commits/004c8453247957e0b7dd4d5326187eec340b7d95, which I hope is less concerning.

I think being able to reliably pick up errors is beyond me - there's such a tangle of things that are potentially valid that I'd prefer it just to fall back to rendering as text.

File: spec/visual/samples/sas (hunk 16 above)
Comment (tomsutch) [reply spec/visual/samples/sas]: Not particularly. SAS generally ignores newlines, it's looking out for semicolons to delimit statements.

File: lib/rouge/lexers/sas.rb [hunk 19]
```diff
@@ -0,0 +1,563 @@
...
+        # Data step definitions
+        rule /(data)(\s+)([\w\d\.]+)/i do
+          groups Keyword, Text, Name::Variable
+        end
+        # Libname definitions
+        rule /(libname)(\s+)([\w\d]+)/i do
+          groups Keyword, Text, Name::Variable
+        end
+
+        rule /\w[\w\d]*/ do |m|
```
Comment (ashmaroli) [inline lib/rouge/lexers/sas.rb]: I see many uses of including both `\w` and `\d` in the same character set. That is unnecessary as `\d` is a subset of `\w`
`/\w+/` will match `foo` as well as `f00`.

You may test this by invoking `irb` with the `-w` flag:
```sh
irb -w
```

File: spec/visual/samples/sas (hunk 12 above)
Comment (tomsutch) [reply spec/visual/samples/sas]: I've now added a link to the licence https://github.com/rouge-ruby/rouge/pull/1107/commits/3191ee40b262959258188a1c4661249f0daf3520 so I think this covers the bases as suggested in [CC's best practices document](https://wiki.creativecommons.org/wiki/best_practices_for_attribution).

File: lib/rouge/lexers/sas.rb (hunk 18 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks, now changed.

File: lib/rouge/lexers/sas.rb (hunk 19 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thank you. I have changed in commit https://github.com/rouge-ruby/rouge/pull/1107/commits/b1f3ae3e67387bc6750aa1eac34228c3a6dd5d04.

File: lib/rouge/demos/sas [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+        x2 = rand("Binomial", 0.5, 100);
+        output;
+    end;
+run;
+
+proc means data=sashelp.class;
+    class sex;
+    var height weight;
+    output out = mean_by_sex;
+run;
```
[the remark this answers is not in the document]
Comment (tomsutch) [reply lib/rouge/demos/sas]: Thanks; I've added a newline in ac2f4bd053e6ce89d97485fd4cf34e807193bbd4

File: lib/rouge/lexers/sas.rb (hunk 9 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: I think so - here is the [SAS documentation about operators](https://documentation.sas.com/?docsetId=lrcon&docsetTarget=p00iah2thp63bmn1lt20esag14lh.htm&docsetVersion=9.4&locale=en) which I was working from.

* line 427: matching ** (exponentiation) or \ * + -
* line 428: matching ^=, ¬=, ~= as variants of 'not equal to' (see the Comparison Operators table in the linked page), as well as >=, >, <=, <, =. And then you can add a colon modifier to any of these to compare only a specified prefix :thinking: 
* line 429: mnemonic equivalents, as shown in the 'Comparison Operators' table in the linked page
* line 430: these are in the Logical Operators table in the linked page. There are lots of NOT alternatives!

File: lib/rouge/lexers/sas.rb (hunk 11 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: I was going for an instance variable so that when in `state :proc` we would know which PROC we were in and could retrieve the relevant keywords for that. Happy to amend if there's a better way of doing it!

**Comment by ashmaroli** [comment]
You'll need to resolve all offenses reported by RuboCop:
https://travis-ci.org/rouge-ruby/rouge/jobs/554691134

File: lib/rouge/lexers/sas.rb (hunk 11 above)
Comment (pyrmont) [reply lib/rouge/lexers/sas.rb]: Ah, I see what you mean. My initial reaction is that this kind of checking isn't really the type of thing to do in the lexer, you could just groups all the keywords together and if that means some false positives, it shouldn't be the end of the world.

Now that you've gone and done all the hard work, though, I'm happy to leave as-is.

Comment (pyrmont) [reply lib/rouge/lexers/sas.rb]: Oh, one thing, though is that you shouldn't need to split `m[0]` since you're capturing the terms separately. You should be able to get the keyword from `m[3]`.

**Comment by pyrmont** [comment]
@tomsutch Any luck fixing the outstanding items in this PR?

File: lib/rouge/lexers/sas.rb [hunk 21]
```diff
@@ -0,0 +1,563 @@
...
+        end
+
+
+        # operators (Language Reference p96)
+        rule %r(\*\*|[\*/\+-]), Operator
+        rule /[^¬~]?=:?|[<>]=?:?/, Operator
+        rule /\b(eq|ne|gt|lt|ge|le|in)\b/i, Operator::Word
+        rule /[&|!¦¬∘~]/, Operator
+        rule /\b(and|or|not)\b/i, Operator::Word
+        rule /[<>|><]/, Operator # min/max
```
[the remark this answers is not in the document]
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks; committed in 10d2c0a

File: lib/rouge/lexers/sas.rb [hunk 22]
```diff
@@ -0,0 +1,563 @@
...
+        rule /\?/, Str::Symbol # used for format modifiers
+        
+        rule /[^\s]+/, Text # Fallback for anything we haven't matched so far
+      end
+
+      state :root do
+        mixin :basics
+
+        # PROC definitions
+        rule %r((proc)(\s+)(\w+))ix do |m|
```
[the remark this answers is not in the document]
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks; committed in 10d2c0a

File: lib/rouge/lexers/sas.rb (hunk 11 above)
Comment (tomsutch) [reply lib/rouge/lexers/sas.rb]: Thanks; fixed in https://github.com/rouge-ruby/rouge/pull/1107/commits/f0a0871123d7062290d9734d26a0fd66c6fd4fc1 . For simplicity I also changed the keys of the hash to upper case -- I was previously doing `downcase` then `upcase` again!

**Comment by ashmaroli** [comment]
@tomsutch I hate to inform you but looks like Travis CI failed with some linting errors. Please fix.
Thanks.

**Comment by pyrmont** [comment]
@tomsutch Before pushing, it's a good idea to run the following:

```shell
rake && bundle exec rubucop bin lib spec tasks
```

If that runs without error then your code should pass Travis as well.

**Comment by tomsutch** [comment]
> You'll need to resolve all offenses reported by RuboCop:
> https://travis-ci.org/rouge-ruby/rouge/jobs/554691134

Thanks - the latest build is https://travis-ci.org/rouge-ruby/rouge/jobs/559860704
It consists entirely of 'offenses' like this

```
Lint/AmbiguousRegexpLiteral: Ambiguous regexp literal. Parenthesize the method arguments if it's surely a regexp literal, or add a whitespace to the right of the / if it should be a division.
```
There's one offense for every instance of `rule /.../`

I'm certainly happy to do what's needed to fix this, but I don't understand why these are happening when other lexers contain similar rules in similar formats and don't get picked up by Rubocop. Do you know why this might be?

**Comment by ashmaroli** [comment]
The offense is simply `rule /.../`. Other lexers may have it like `rule %r/.../` or `rule(/.../)`

**Comment by pyrmont** [comment]
@tomsutch The explanation given by RuboCop is brief but, essentially, RuboCop flags the use of literal regular expressions of the form `/.../` in method arguments as ambiguous because it's not clear whether the initial `/` represents the beginning of a regular expression or a method called `/`.

Other lexers avoid this warning by either using a sigil to mark the beginning of the regex (`%r` so that it becomes `%r/.../`) or wrapping the parameters inside parentheses.

**Comment by tomsutch** [comment]
> @tomsutch The explanation given by RuboCop is brief but, essentially, RuboCop flags the use of literal regular expressions of the form `/.../` in method arguments as ambiguous because it's not clear whether the initial `/` represents the beginning of a regular expression or a method called `/`.
> 
> Other lexers avoid this warning by either using a sigil to mark the beginning of the regex (`%r` so that it becomes `%r/.../`) or wrapping the parameters inside parentheses.

Ah, thanks. Now I see that the the other lexers I was referring to when putting this together have also been changed accordingly via https://github.com/rouge-ruby/rouge/pull/1180. I'll fix this now.

**Comment by tomsutch** [comment]
I think I've now addressed all the feedback, if you're happy with my response to your request @pyrmont?
Thanks everyone for your patience!

**Comment by pyrmont** [comment]
@tomsutch Thanks for the contribution to Rouge :) Another language that's highlighted! 🎉
